### PR TITLE
Adding reactor netty read and write timeout

### PR DESCRIPTION
**Pull request by simplynaveen20**
Closes https://github.com/Azure/azure-sdk-for-java/issues/13696

Note : Updating Mockito version , and adding opt in mocking final class, [check this](https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods)

Todo - Will do perf regression testing

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/pom.xml [hunk 1]
```diff
@@ -231,7 +231,7 @@ Licensed under the MIT License.
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-core</artifactId>
-      <version>1.10.19</version> <!-- {x-version-update;cosmos_org.mockito:mockito-core;external_dependency} -->
+      <version>3.4.6</version> <!-- {x-version-update;cosmos_org.mockito:mockito-core;external_dependency} -->
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/pom.xml:234]: Though one issue here - you updated the dependency - but didn't use it correctly here. 

There are 2 things you need to do. 

In this file - https://github.com/Azure/azure-sdk-for-java/blob/master/eng/versioning/external_dependencies.txt
you need to remove this line - https://github.com/Azure/azure-sdk-for-java/blob/master/eng/versioning/external_dependencies.txt#L244

And in your azure-cosmos/pom.xml - you need to use the mockito-core version mentioned here - https://github.com/Azure/azure-sdk-for-java/blob/master/eng/versioning/external_dependencies.txt#L216
And update the tag to below . 

```
<version>3.0.0</version> <!-- {x-version-update;org.mockito:mockito-core;external_dependency} -->
```

If this doesn't make sense, let me know and I will explain you how this external_dependencies work in central SDK.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyClientTest.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            .buildAsyncClient();
+    }
+
+    @DataProvider(name = "cosmosClients")
+    public Object[][] cosmosClients() {
+        return new Object[][]{
+            {gatewayClient}, {directClient}};
+    }
+
+    @Test(groups = {"emulator"}, dataProvider = "cosmosClients", timeOut = SETUP_TIMEOUT)
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/http/ReactorNettyClientTest.java:66]:
```suggestion
    @Test(groups = {"emulator"}, dataProvider = "cosmosClients", timeOut = TIMEOUT)
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java [hunk 3]
```diff
@@ -82,7 +85,9 @@ private void configureChannelPipelineHandlers() {
             //  By default, keep alive is enabled on http client
             tcpClient = tcpClient.option(ChannelOption.CONNECT_TIMEOUT_MILLIS,
                 (int) configs.getConnectionAcquireTimeout().toMillis());
-
+            tcpClient = tcpClient.doOnConnected(conn -> {
+                conn.addHandlerLast(new WriteTimeoutHandler(this.httpClientConfig.getRequestTimeout().toMillis(), TimeUnit.MILLISECONDS));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:89]: please make sure for both Gateway and DirectHttps  io request timeout from public surface api is correctly plumbed through to `httpHttpConfig.getRequestTimeout()`

**Comment by kushagraThapar** [comment]
@simplynaveen20  @moderakh - I found out that this `ReadTimeoutHandler` and `WriteTimeoutHandler` will not work well for our case here - because they interfere with idle connections and will cause connections to close even when they are ready to be re-used by other requests. 

Please see this issue for more detail: https://github.com/reactor/reactor-netty/issues/1159

I think we should wait for reactor-netty 0.9.11.RELEASE to properly support this as it will be introduced as a new API in their code.

**Pull request closed without merging**